catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,129 @@
+
+    private int ceilingIndex;
+
+    private int endIndex;
+
+    private int maxResults;
+
+    private int indexOffset;
+
+    public QueryResultPaginator(CatalogFramework catalogFramework, QueryRequestImpl queryRequest) {
+        this.catalogFramework = catalogFramework;
+        this.queryRequest = queryRequest;
+
+        // maxResults == total number of results to return to client
+        maxResults = queryRequest.getQuery().getPageSize();
+
+        indexOffset = 1;
+        endIndex = 100000000;
+        ceilingIndex = 100000000;
+    }
+
+    public QueryResultPaginator(CatalogFramework catalogFramework, QueryRequestImpl queryRequest,
+            int indexOffset, int endIndex, int maxResults, int ceilingIndex) throws Exception {
+
+        this.catalogFramework = catalogFramework;
+        this.queryRequest = queryRequest;
+        this.indexOffset = indexOffset;
+        this.endIndex = endIndex;
+        this.maxResults = maxResults;
+        this.ceilingIndex = ceilingIndex;
+
+        if (indexOffset < 0 || indexOffset > ceilingIndex || indexOffset > endIndex) {
+            throw new Exception();
+        }
+    }
+
+    // TODO: 6/14/17 Determine how this is used and best way to check
+    //    public boolean hasNext() {
+    //        boolean hasNext;
+    //
+    //        if (maxResults > 0) {
+    //            hasNext = true;
+    //        } else {
+    //            hasNext = false;
+    //        }
+    //
+    //        return hasNext;
+    //    }
+
+    public List<Result> nextPage() {
+        List<Result> resultList = new ArrayList<>();
+
+        final int originalIndexOffset = indexOffset;
+
+        while (indexOffset < endIndex && resultList.size() < maxResults
+                && indexOffset < ceilingIndex) {
+
+            resultList.addAll(queryCatalogFramework());
+            resultList = trimListToMaxResults(resultList,
+                    maxResults,
+                    ceilingIndex,
+                    originalIndexOffset);
+
+            indexOffset = originalIndexOffset + resultList.size();
+        }
+        return resultList;
+    }
+
+    private List<Result> queryCatalogFramework() {
+        try {
+            Map<String, Serializable> props = new HashMap<>();
+            // Avoid caching all iteratorResults while dumping with native query mode
+            props.put("mode", "native");
+            response = catalogFramework.query(queryRequest);
+        } catch (UnsupportedQueryException | SourceUnavailableException | FederationException e) {
+            throw new CatalogRuntimeException(e);
+        }
+        List<Result> resultList = response.getResults();
+        return resultList;
+    }
+
+    private List<Result> trimListToMaxResults(List<Result> list, int maxResults, int ceilingIndex,
+            int originalIndexOffset) {
+
+        // TODO: 6/15/17 Change to feedback on number of results in last query
+        if (list.size() > maxResults) {
```

Comment:
What is the performance impact of dropping the unpaged results? Looks like we make a full query every time we want a new page and drop everything that isn't in the range.